src/KeyringController.ts

```diff
@@ -442,68 +544,141 @@ class KeyringController extends EventEmitter {
+    }
+
+    return keyring.signTypedData(
+      address,
+      msgParams.data as Eip1024EncryptedData,
+      opts,
+    );
   }
 
   /**
    * Gets the app key address for the given Ethereum address and origin.
    *
-   * @param {string} _address - The Ethereum address for the app key.
-   * @param {string} origin - The origin for the app key.
-   * @returns {string} The app key address.
+   * @param _address - The Ethereum address for the app key.
+   * @param origin - The origin for the app key.
+   * @returns The app key address.
    */
-  async getAppKeyAddress(_address, origin) {
-    const address = normalizeAddress(_address);
+  async getAppKeyAddress(_address: string, origin: string): Promise<string> {
+    const address = normalizeAddress(_address) as Hex;
     const keyring = await this.getKeyringForAccount(address);
+    if (!keyring.getAppKeyAddress) {
+      throw new Error(
+        `KeyringController - The keyring for address ${_address} does not support the method getAppKeyAddress.`,
+      );
+    }
+
     return keyring.getAppKeyAddress(address, origin);
   }
 
   /**
    * Exports an app key private key for the given Ethereum address and origin.
    *
-   * @param {string} _address - The Ethereum address for the app key.
-   * @param {string} origin - The origin for the app key.
-   * @returns {string} The app key private key.
+   * @param _address - The Ethereum address for the app key.
+   * @param origin - The origin for the app key.
+   * @returns The app key private key.
    */
-  async exportAppKeyForAddress(_address, origin) {
-    const address = normalizeAddress(_address);
+  async exportAppKeyForAddress(
+    _address: string,
+    origin: string,
+  ): Promise<string> {
+    const address = normalizeAddress(_address) as Hex;
     const keyring = await this.getKeyringForAccount(address);
     // The "in" operator is typically restricted because it also checks inherited properties,
     // which can be unexpected for plain objects. We're allowing it here because `keyring` is not
     // a plain object, and we explicitly want to include inherited methods in this check.
     // eslint-disable-next-line no-restricted-syntax
-    if (!('exportAccount' in keyring)) {
+    if (!keyring.exportAccount) {
       throw new Error(
-        `The keyring for address ${_address} does not support exporting.`,
+        `KeyringController - The keyring for address ${_address} does not support exporting.`,
       );
     }
     return keyring.exportAccount(address, { withAppKeyOrigin: origin });
   }
 
-  //
-  // PRIVATE METHODS
-  //
+  /**
+   * Restore Keyring
+   *
+   * Attempts to initialize a new keyring from the provided serialized payload.
+   * On success, updates the memStore keyrings and returns the resulting
+   * keyring instance.
+   *
+   * @param serialized - The serialized keyring.
+   * @returns The deserialized keyring.
+   */
+  async restoreKeyring(serialized: any): Promise<Keyring<State> | undefined> {
```

Comment:
Is it possible to use `Json` or something here instead of `any`? Or is that not possible because this could theoretically come from anywhere?

Reply:
Good question. Using `Json` here would make sense to me.

It looks like this comes from the decryptor, which types this as `unknown`. But it should be guaranteed to be JSON in practice. See here for extension: https://github.com/MetaMask/browser-passworder/blob/main/src/index.ts#L176 and mobile: https://github.com/MetaMask/metamask-mobile/blob/main/app/core/Encryptor.js#L75

Reply:
I have defined the new data type `SerializedKeyring` in [ce7d126](https://github.com/MetaMask/KeyringController/pull/202/commits/ce7d126375790bd07789a82f6c9609f0a2eb9813). Let me know what you think